dbms/src/Processors/Formats/Impl/ConstantExpressionTemplate.cpp

```diff
@@ -0,0 +1,538 @@
+        if (settings.values.accurate_types_of_literals && !structure->special_parser[cur_column].useDefaultParser())
+        {
+            if (!parseLiteralAndAssertType(istr, type.get(), cur_column))
+                return false;
+        }
+        else
+            type->deserializeAsTextQuoted(*columns[cur_column], istr, settings);
+
+        ++cur_column;
+    }
+    while (cur_token < structure->tokens.size())
+    {
+        skipWhitespaceIfAny(istr);
+        if (!checkString(structure->tokens[cur_token++], istr))
+            return false;
+    }
+
+    return true;
+}
+
+bool ConstantExpressionTemplate::parseLiteralAndAssertType(ReadBuffer & istr, const IDataType * complex_type, size_t column_idx)
+{
+    using Type = Field::Types::Which;
+
+    /// TODO in case of type mismatch return some hints to deduce new template faster
+    if (istr.eof())
+        return false;
+
+    SpecialParserType type_info = structure->special_parser[column_idx];
+
+    /// If literal does not fit entirely in the buffer, parsing error will happen.
+    /// However, it's possible to deduce new template (or use template from cache) after error like it was template mismatch.
+
+    if (type_info.is_array)
+    {
+        /// TODO faster way to check types without using Parsers
+        ParserArrayOfLiterals parser_array;
+        Tokens tokens_number(istr.position(), istr.buffer().end());
+        IParser::Pos iterator(tokens_number);
+        Expected expected;
+        ASTPtr ast;
+
+        if (!parser_array.parse(iterator, ast, expected))
+            return false;
+        istr.position() = const_cast<char *>(iterator->begin);
+
+        const Field & array = ast->as<ASTLiteral &>().value;
+        auto array_type = applyVisitor(FieldToDataType(), array);
+        auto nested_type = dynamic_cast<const DataTypeArray &>(*array_type).getNestedType();
+        if (type_info.is_nullable)
+            if (auto nullable = dynamic_cast<const DataTypeNullable *>(nested_type.get()))
+                nested_type = nullable->getNestedType();
+
+        WhichDataType nested_type_info(nested_type);
+        if ((nested_type_info.isNativeUInt() && type_info.nested_type == Type::UInt64) ||
+            (nested_type_info.isNativeInt()  && type_info.nested_type == Type::Int64)  ||
+            (nested_type_info.isFloat64()    && type_info.nested_type == Type::Float64))
+        {
+            Field array_same_types = convertFieldToType(array, *complex_type, nullptr);
+            columns[column_idx]->insert(array_same_types);
+            return true;
+        }
+        return false;
+    }
+    else
+    {
+        Field number;
+        if (type_info.is_nullable && 4 <= istr.available() && 0 == strncasecmp(istr.position(), "NULL", 4))
+            istr.position() += 4;
+        else
+        {
+            /// ParserNumber::parse(...) is slow because of using ASTPtr, Expected and Tokens, which is't needed here
```

Comment:
How much slower is it?

Reply:
As I remember, `strtod` takes only about 5% of `ParserNumber::parse`, so it's better to use `strtod` and `strtoull` directly here